lxd/instance/drivers/driver_lxc.go

```diff
@@ -1301,6 +1301,60 @@ func (d *lxc) initLXC(config bool) error {
 	return nil
 }
 
+var idmappedStorageMap map[unix.Fsid]idmap.IdmapStorageType = map[unix.Fsid]idmap.IdmapStorageType{}
+var idmappedStorageMapLock sync.Mutex
+
+// IdmappedStorage determines if the container can use idmapped mounts or shiftfs
+func (d *lxc) IdmappedStorage(path string) idmap.IdmapStorageType {
+	if path == "" {
+		path = d.RootfsPath()
+	}
+
+	buf := &unix.Statfs_t{}
+
+	if !d.state.OS.LXCFeatures["idmapped_mounts"] {
+		if d.state.OS.Shiftfs {
+			return idmap.IdmapStorageShiftfs
+		}
+
+		return idmap.IdmapStorageNone
+	}
+
+	err := unix.Statfs(path, buf)
+	if err != nil {
+		return idmap.IdmapStorageNone
+	}
+
+	idmappedStorageMapLock.Lock()
+	val, ok := idmappedStorageMap[buf.Fsid]
+	idmappedStorageMapLock.Unlock()
+	if ok {
+		return val
+	}
+
+	canIdmap := idmap.CanIdmapMount(path)
+	idmappedStorageMapLock.Lock()
+
+	// Recheck in case someone was faster than we were.
+	val, ok = idmappedStorageMap[buf.Fsid]
+	if ok {
+		return val
```

Comment:
@brauner this is leaving the lock in place, consider using `defer idmappedStorageMapLock.Unlock()` after its locked on line 1336.

Reply:
Done.